File: go-version/leetcode/LC2962/count_subarrays.go
package LC2962

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/4/29 下午11:13
 * @FilePath: count_subarrays
 * @Description: 滑动窗口，越长越合法
 */

func countSubarrays(nums []int, k int) int64 {
	var ans int64
	var cntMax, left, mx int
	for _, num := range nums {
		mx = max(mx, num)
	}

	for i := range nums {
		if nums[i] == mx {
			cntMax++
		}
		for cntMax >= k {
			if nums[left] == mx {
				cntMax--
			}
			left++
		}
		ans += int64(left)
	}
	return ans
}

File: python-version/LC2962/count_subarrays.py
class Solution:
    @staticmethod
    def countSubarrays(nums: list[int], k: int) -> int:
        mx = max(nums)
        left, ans, cnt_max = 0, 0, 0
        for num in nums:
            if num == mx:
                cnt_max += 1
            while cnt_max >= k:
                if nums[left] == mx:
                    cnt_max -= 1
                left += 1
            ans += left
        return ans
